docs/proposals/hpa-v2.md

```diff
@@ -0,0 +1,228 @@
+
+The current Horizontal Pod Autoscaler object only has support for CPU as
+a percentage of requested CPU.  While this is certainly a common case, one
+of the most frequently sought-after features for the HPA is the ability to
+scale on different metrics (be they custom metrics, memory, etc).
+
+The current HPA controller supports targeting "custom" metrics (metrics
+with a name prefixed with "custom/") via an annotation, but this is
+suboptimal for a number of reasons: it does not allow for arbitrary
+"non-custom" metrics (e.g. memory), it does not allow for metrics
+describing other objects (e.g. scaling based on metrics on services), and
+carries the various downsides of annotations (not be typed/validated,
+being hard for a user to hand-construct, etc).
+
+Object Design
+-------------
+
+### Requirements ###
+
+This proposal describes a new version of the Horizontal Pod Autoscaler
+object with the following requirements kept in mind:
+
+1. The HPA should continue to support scaling based on percentage of CPU
+   request
+
+2. The HPA should support scaling on arbitrary metrics associated with
+   pods
+
+3. The HPA should support scaling on arbitrary metrics associated with
+   other Kubernetes objects in the same namespace as the HPA (and the
+   namespace itself)
+
+4. The HPA should make scaling on multiple metrics in a single HPA
+   possible and explicit (splitting metrics across multiple HPAs leads to
+   the possibility of fighting between HPAs)
+
+### Specification ###
+
+```go
+type HorizontalPodAutoscalerSpec struct {
+    // the target scalable object to autoscale
+    ScaleTargetRef CrossVersionObjectReference `json:"scaleTargetRef"`
+
+    // the minimum number of replicas to which the autoscaler may scale
+    // (optional, defaults to 1)
+    MinReplicas *int32 `json:"minReplicas,omitempty"`
+    // the maximum number of replicas to which the autoscaler may scale
+    MaxReplicas int32 `json:"maxReplicas"`
+
+    // the target CPU percentage (optional)
+    TargetCPUUtilizationPercentage `json:"targetCPUUtilizationPercentage,omitemtpy"`
+
+    // the non-CPU percentage metrics (optional)
+    Metrics []MetricSpec `json:"metrics,omitempty"`
+}
+
+// a specification for how to scale based on a single metric.
+// only one of the fields should be set at once (similarly to
+// volumes, env var types, etc)
+type MetricSpec struct {
+    // metric describing a single Kubernetes object
+    Object *ObjectMetricSource `json:"object,omitempty"`
+
+    // metric describing pods in the scale target
+    Pods *PodsMetricSource `json:"pods,omitemtpy"`
+}
+
+// a metric describing a Kubernetes object
+type ObjectMetricSource struct {
+    // the described Kubernetes object
+    Target CrossVersionObjectReference `json:"target"`
+
+    // the name of the metric in question
+    MetricName string `json:"metricName"`
+
+    // the target value of the metric (as a quantity)
+    TargetValue resource.Quantity `json:"targetValue"`
```

Comment:
Perhaps I'm misunderstanding you here, but I don't think it should (or can be) our responsibility to decide which metrics a user can scale on.  We (as Kubernetes) authors can't really know which metrics might vary proportionally to load for a given application.  As a somewhat "silly" example, error rate might vary proportionally to load if the bulk of the errors are "max connections queued in server" (but you probably shouldn't autoscale on error rate, even then).